vta/hardware/chisel/src/main/scala/Compute.scala

```diff
@@ -0,0 +1,364 @@
+    .elsewhen (push_prev_dep) { finish_wrap := push_prev_dep_ready && busy }
+    .elsewhen (push_next_dep) { finish_wrap := push_next_dep_ready && busy }
+    .otherwise { finish_wrap := false.B }
+  } .otherwise { finish_wrap := false.B }
+  when (uop_cntr_wrap || acc_cntr_wrap || out_cntr_wrap || finish_wrap) {
+    when (push_prev_dep || push_next_dep) {
+      state := s_PUSH
+    } .otherwise {
+      state := s_DONE
+    }
+  }
+
+  // dependency queue processing
+  when (busy && (!pop_prev_dep_ready && !pop_next_dep_ready) && (pop_prev_dep || pop_next_dep)) { state := s_DUMP }
+  when (dump && ( pop_prev_dep_ready ||  pop_next_dep_ready)) { state := s_BUSY }
+  when (push && (push_prev_dep_ready || push_next_dep_ready)) { state := s_DONE }
+
+  // dependency queue processing
+  io.l2g_dep_queue.ready := pop_prev_dep_ready && dump
+  io.s2g_dep_queue.ready := pop_next_dep_ready && dump
+  io.l2g_dep_queue.data <> DontCare
+  io.s2g_dep_queue.data <> DontCare
+  when (pop_prev_dep && dump && io.l2g_dep_queue.valid) {
+    pop_prev_dep_ready := true.B
+  }
+  when (pop_next_dep && dump && io.s2g_dep_queue.valid) {
+    pop_next_dep_ready := true.B
+  }
+  io.g2l_dep_queue.data := 1.U
+  io.g2s_dep_queue.data := 1.U
+  io.g2l_dep_queue.valid := push_prev_dep_valid
+  io.g2s_dep_queue.valid := push_next_dep_valid
+  when (push_prev_dep_valid && io.g2l_dep_queue.ready && push) {
+    push_prev_dep_ready := true.B
+  }
+  when (push_next_dep_valid && io.g2s_dep_queue.ready && push) {
+    push_next_dep_ready := true.B
+  }
+
+  // setup counters
+  when (uops_read && !uop_cntr_wait && busy && uop_cntr_val < uop_cntr_max) {
+    uop_cntr_val := uop_cntr_val + 1.U
+  }
+  when (biases_read && !acc_cntr_wait && busy && acc_cntr_val < acc_cntr_max) {
+    acc_cntr_val := acc_cntr_val + 1.U
+  }
+  // when (out_mem_write && !out_cntr_wait && busy && out_cntr_val < out_cntr_max) {
+  //   out_cntr_val := out_cntr_val + 1.U
+  // }
+  when (out_mem_write && busy && out_cntr_val < out_cntr_max) {
+    out_cntr_val := out_cntr_val + 1.U
+  }
+
+  // fetch instruction
+  when (gemm_queue_ready) {
+    insn := io.gemm_queue.data
+    uop_cntr_val := 0.U
+    acc_cntr_val := 0.U
+    out_cntr_val := 0.U
+    pop_prev_dep_ready := false.B
+    pop_next_dep_ready := false.B
+    push_prev_dep_ready := false.B
+    push_next_dep_ready := false.B
+    state := s_BUSY
+  } .otherwise {
+    insn := insn
+  }
+  gemm_queue_ready := io.gemm_queue.valid && (idle || done)
+  io.gemm_queue.ready := gemm_queue_ready
+  when (gemm_queue_ready) { gemm_queue_ready := false.B }
+
+  // response to done signal
+  io.done.waitrequest := 0.U
+  io.done.write <> DontCare
+  io.done.writedata <> DontCare
+  io.done.readdata := opcode_finish_en.asUInt
+
+  // fetch uops
+  val uop_mem_write_en = (opcode_load_en && (memory_type === mem_id_uop.U))
+  val uop_dram_addr = (dram_idx + uop_cntr_val) << log2Ceil(128 / 8).U
```

Comment:
strictly speaking, you don't need to `.U` on `log2Ceil`, but the constant gets inlined during compilation anyway (also below)